Enable nxp MICFIL driver and Dialog DA7212 driver

ZhaoxiangJin opened the pull request:
1. Enabled DA7212 driver
2. Tested da7212 with i2s_codec example (CONFIG_USE_DMIC=n, you can hear the sine wave from the codec.)
2. Enabled MICFIL driver
3. Tested MICFIL with i2s_codec example (CONFIG_USE_DMIC=y, you can hear the voice played from the on-board micphone)

According to the reviewer's suggestion, this PR has been split into the following PRs:

1. [x] https://github.com/zephyrproject-rtos/zephyr/pull/96975
2. [x] https://github.com/zephyrproject-rtos/zephyr/pull/96976
3. [x] https://github.com/zephyrproject-rtos/zephyr/pull/96978
4. [x] https://github.com/zephyrproject-rtos/zephyr/pull/96980

TODO: There may be some problems with the NXP port driver that cause the MICFIL CLOCK and DATA pinmux to not be configured successfully. The current workaround is to [write to the registers directly](https://github.com/zephyrproject-rtos/zephyr/pull/96367/commits/aa297baed416334c275bba127de34bf30271a65f#diff-f422348e5214c65a87e7e13db10c415503abc4379699ea7abe9119ec9b9d9270R351). The root cause will be found and solved later.

decsny commented on dts/bindings/audio/nxp,pdm.yaml:
```diff
@@ -0,0 +1,87 @@
+
+description: NXP PDM (Pulse Density Microphone) controller
+
+compatible: "nxp,pdm"
```
is there a more specific name to give this? I think nxp has multiple PDM controller? @DerekSnell

ZhaoxiangJin replied:
Rename to MICFIL in the latest update

VitekST commented on drivers/audio/dmic_nxp_pdm.c:
```diff
@@ -0,0 +1,550 @@
+	for (uint16_t frame = 0; frame < frames_to_read; frame++) {
+		for (uint8_t chan = 0; chan < data->channels; chan++) {
+			uint8_t hw_ch = data->hw_chan[chan];
+			/* Take high 24 bits: shift down by 8. Low 8 bits are zero in FIFO */
```
AFAIK, 24 bit PCM samples can be quite problematic when dealing with them later, for example not all I2S codecs support I2S traffic bearing such samples. 32 bits is usually the "greatest common divisor". I suggest that you either just copy from FIFO as it is (thus forcing 32 bits upon the user) or implement some logic that allows the user to select a bit width (16, 24 and 32 perhaps) and which truncates the samples as needed.

ZhaoxiangJin committed:
audio: dmic: Format dmic.h

Remove whitespaces to make code easier to read
and follow the common sense coding style.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
drivers: audio: Add dialog7212 driver

Add dialog7212 driver.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
boards: frdm_mcxn236: Enable da7212 on frdm_mcxn236

Enable da7212 on frdm_mcxn236

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
drivers: audio: Add NXP MICFIL driver

Add NXP MICFIL driver base DMIC device driver model.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
drivers: mcux_syscon: enable MICFIL clock control

Enable MICFIL clock control for NXP MCUX SoCs.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
board: frdm_mcxn236: Enable MICFIL on frdm_mcxn236

1. Enable MICFIL on frdm_mcxn236 board.
2. MICFIL CLOCK and DATA Pins are conflict with
flexcomm0_lpuart pins, so change flexcomm0_lpuart
pins to 'FC0_P2_PIO0_6' and 'FC0_P3_PIO0_7'.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
samples: i2s_codec: Refine sample for samll RAM platforms

- Configure BLOCK count by using kconfig option
'EXTRA_BLOCKS' instad of fixed value. This allows user to
customize the RAM usage based on their platform.
- Keep the PCM data in flash to avoid large RAM usage.
- Refine the sample to get the SAMPLE_BIT_WIDTH
and BYTES_PER_SAMPLE from Kconfig options.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
samples: i2s_codec: enable i2s_codec samples for frdm_mcxn236

Enable i2s_codec samples for frdm_mcxn236.
With 'CONFIG_USE_DMIC=n', you can hear the sine wave
from the codec.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
samples: i2s_codec: Enable nxp micfil test

Enable nxp micfil test with 'CONFIG_USE_DMIC=y',
you can hear the voice played from the on-board micphone.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
dts: bindings: add dialog vendor prefix

Add `dialog` for Dialog Semiconductor.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

VitekST commented on drivers/audio/dmic_nxp_pdm.c:
```diff
@@ -0,0 +1,550 @@
+					index, &pdm_idx, &lr);
+
+		/* Map (pdm_idx, lr) to hardware DATACH index: 2*pdm_idx + (R?1:0). */
+		uint8_t hw_chan = (uint8_t)(2U * pdm_idx + (lr == PDM_CHAN_RIGHT ? 1U : 0U));
```
@ZhaoxiangJin On a second glance, I think you have a point. 

The mentioned constraint isn't worded with needed clarity, but the `dmic_nrfx_pdm.c` driver from Nordic seems to be written in such a way that the left/right distinction is indeed relevant - what it calls a "channel" doesn't map to a singular PCM stream, but rather to a pair of those, which are then distinguished with the `enum pdm_lr` values. Each hardware channel index corresponds conceptually to a PDM data line (of course, carrying two channels, because of the DDR signalling).

...so I think that the way the `nxp,dmic` driver got modified does indeed deviate from the API as it was intended to be used and that change should've really been an API change with a proper RFC.

Now... not realising this, I authored some changes in the `i2s_codec` demo that build a 1:1 map of channels. These changes were merged, enabling the demo on `mimxrt685_evk` and `mimxrt595_evk` demos. This constitutes bit of a sticky situation, as there are now drivers with differing interpretations of the `dmic` API. The `dmic_api` is written against the original one, `i2s_codec` is written against the modified one.

I don't feel I am the right person to make decisions such as these, but what I would perhaps do is stick to the API as it is defined at the moment and just have the `dmic_api` test functional, with the `i2s_codec` sample being broken for that target. I can revert that particular commit if an RFC doesn't come through.

dbaluta reviewed (changes requested):
@ZhaoxiangJin this driver looks oddly similar with this one:

https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/dai/nxp/micfil/micfil.c

Can't you use this? Also, you haven't answered to any of my comments.

ZhaoxiangJin commented on include/zephyr/audio/dmic.h:
```diff
@@ -81,13 +81,13 @@ struct pdm_io_cfg {
 	 *@{
 	 */
 	/** Minimum clock frequency supported by the mic */
```
Update, hope you have see the change

VitekST replied:
@ZhaoxiangJin Changes in this file are just whitespace changes, from what I see. I don't think you're supposed to reformat code (albeit rightfully) that you haven't changed in any other way.

ZhaoxiangJin replied:
Personally, I think consistency in indentation style is an important part of the code, so this commit is necessary.

ZhaoxiangJin commented on drivers/audio/Kconfig.dmic_nxp_micfil:
```diff
@@ -0,0 +1,22 @@
+	select CLOCK_CONTROL
+	help
+	  Enable NXP PDM driver.
+
```
renamed the driver to micfil, please review

ZhaoxiangJin commented:
> @ZhaoxiangJin this driver looks oddly similar with this one:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/dai/nxp/micfil/micfil.c
> 
> Can't you use this? Also, you haven't answered to any of my comments. 


I think DAI and DMIC should be different device driver models. Of course, I haven't closely examined exactly what kind of model DAI is.

VitekST commented:
> > @dbaluta That `nxp,dai-micfil` driver handles only the 1st PDM channel, as it exposes only the 1st (`DATACH0`) PCM FIFO. In addition, there's no sample for the `dai` API (the only one is, well, SOF) and it pushes the need to handle data transfers onto the application developer.
> 
> Actually, it handles 1-8 channels with SDMA scripts taking care of copying the data starting at DATACH0 + current_channel. I tested it with SOF and 8 channels.
> 
> I have no problem on you using the I2S interface. Looks like it's there for longer time then DAI interface. I want to understand if we can avoid duplicating the code and having two drivers for the same IP (not clear though it is the same IP) as the naming is confusing and havent seen yet the datasheet for this PRs IPs/

Okay, noted. Yeah, that's trivial, the FIFO registers are consecutive.

The API in question is not `i2s`, it's `dmic`. The main difference is that the `dmic` API handles the task of transferring data to a queue of audio buffers - the application developer doesn't have to set up DMA, instead the application consumes buffers. By extension, the API has a facility to remap channels to an arbitrary order, although to which extent is a matter in itself, about which I have started discussing just now.

The motivation is that there are samples for the `dmic` API (`i2s_codec`) as well as tests (`dmic_api`).

About the deduplication - one approach comes to mind. Each audio endpoint could be represented by a driver implementing the `dai` API (with perhaps the extension of being able to expose multiple FIFOs - so that channel remapping can be done), with an overlay driver that would handle DMA and buffer handling, presenting a nice, `i2s`-like API to the user. This would cover both use-cases, having audio that can be used straight as it is in Zephyr and still maintaining compatibility with SOF.

Of course, this travels towards something that can be called "a grand unifying theory of audio on Zephyr". As a result, I don't think that it can be decided just like that in this PR, it's a change that would touch practically all `i2s`, `dmic` and `dai` drivers.

VitekST commented on drivers/audio/dmic_nxp_micfil.c:
```diff
@@ -0,0 +1,630 @@
+	case DMIC_TRIGGER_RELEASE: {
+		/* Check if we are in a state that can be started/released */
+		if (!(data->state == DMIC_STATE_CONFIGURED ||
+			data->state == DMIC_STATE_PAUSED)) {
```
I think that it's also best to use the inequality operator here - use that and replace the disjunction with conjuction.
